### PR TITLE
Chunk the cluster allocation explain response

### PR DESCRIPTION
**Pull request by DiannaHohensee**
Chunking the response can save potentially MBs of heap space.
Leverages ClusterInfo already being chunked (as of #97800).
Chunks the remainder of ClusterAllocationExplanation, and updates
the RestClusterAllocationExplainAction to return a chunked response.

Resolves to issue #97803

### CODE REVIEW

**Comment by DiannaHohensee** [comment]
@DaveCTurner I have several questions about this change.

1) The GitHub issue has `>enhancement` label, but `>non-issue` seems more appropriate? Let me know which is best.

2) There's a lot of wiggle room in what to convert from `toXContent` to `toXContentChunked`. I'm not sure what the rules of thumb are, if there are any. It seems a bit haphazard, at least the way I understand it.

3) Regarding `AbstractChunkedSerializingTestCase.assertChunkCount`, I'm a little lost on how to predict number of chunks expected, aside from just hardcoding whatever the number of chunks happens to be. This brings me back somewhat to (2), whether there's a good way to decide what to put into each chunk.

File: server/src/internalClusterTest/java/org/elasticsearch/action/admin/cluster/allocation/ClusterAllocationExplainIT.java [hunk 1]
```diff
@@ -1239,7 +1242,10 @@ private ClusterAllocationExplanation runExplain(boolean primary, String nodeId,
             XContentBuilder builder = JsonXContent.contentBuilder();
             builder.prettyPrint();
             builder.humanReadable(true);
-            logger.debug("--> explain json output: \n{}", Strings.toString(explanation.toXContent(builder, ToXContent.EMPTY_PARAMS)));
+            logger.debug(
+                "--> explain json output: \n{}",
+                Strings.toString(ChunkedToXContent.wrapAsToXContent(explanation).toXContent(builder, ToXContent.EMPTY_PARAMS))
+            );
```
Comment (DaveCTurner) [inline server/src/internalClusterTest/java/org/elasticsearch/action/admin/cluster/allocation/ClusterAllocationExplainIT.java:1248]: I think we don't need to create our own `XContentBuilder` here any more, this whole block could just be as follows:

```
        if (logger.isDebugEnabled()) {
            logger.debug(
                "--> explain json output: \n{}",
                Strings.toString(explanation, true, true)
            );
        }
```

File: server/src/main/java/org/elasticsearch/action/admin/cluster/allocation/ClusterAllocationExplanation.java [hunk 2]
```diff
@@ -169,48 +174,52 @@ public XContentBuilder toXContent(XContentBuilder builder, Params params) throws
...
-            if (shardAllocationDecision.isDecisionTaken()) {
-                shardAllocationDecision.toXContent(builder, params);
+
+            return builder;
+        }),
+            this.clusterInfo != null
+                ? Iterators.concat(
+                    Iterators.single((builder, p) -> builder.startObject("cluster_info")),
+                    this.clusterInfo.toXContentChunked(params),
+                    Iterators.<ToXContent>single((builder, p) -> builder.endObject())
```
Comment (DaveCTurner) [inline server/src/main/java/org/elasticsearch/action/admin/cluster/allocation/ClusterAllocationExplanation.java:194]: There's some helpers for this case:

```suggestion
                    ChunkedToXContentHelper.startObject("cluster_info"),
                    this.clusterInfo.toXContentChunked(params),
                    ChunkedToXContentHelper.endObject()
```

File: server/src/main/java/org/elasticsearch/cluster/metadata/ShutdownShardMigrationStatus.java [hunk 3]
```diff
@@ -88,7 +89,7 @@ public XContentBuilder toXContent(XContentBuilder builder, Params params) throws
         if (Objects.nonNull(allocationDecision)) {
             builder.startObject(NODE_ALLOCATION_DECISION_KEY);
             {
-                allocationDecision.toXContent(builder, params);
+                ChunkedToXContent.wrapAsToXContent(allocationDecision).toXContent(builder, params);
```
Comment (DaveCTurner) [inline server/src/main/java/org/elasticsearch/cluster/metadata/ShutdownShardMigrationStatus.java:92]: Could we have at least a TODO comment to add chunking here too?

File: server/src/main/java/org/elasticsearch/cluster/routing/allocation/AbstractAllocationDecision.java [hunk 4]
```diff
@@ -127,22 +132,22 @@ public static List<NodeAllocationResult> sortNodeDecisions(List<NodeAllocationRe
...
+            return Collections.emptyIterator();
         }
-        return builder;
+
+        return Iterators.concat(
+            ChunkedToXContentHelper.startArray("node_allocation_decisions"),
+            Iterators.flatMap(
+                nodeDecisions.listIterator(),
+                explanation -> Iterators.single((builder, params) -> explanation.toXContent(builder, params))
+            ),
```
Comment (DaveCTurner) [inline server/src/main/java/org/elasticsearch/cluster/routing/allocation/AbstractAllocationDecision.java:148]: `flatMap`/`single` is a no-op:

```suggestion
            nodeDecisions.iterator(),
```

File: server/src/main/java/org/elasticsearch/common/xcontent/ChunkedToXContentHelper.java [hunk 5]
```diff
@@ -82,6 +82,12 @@ private static <T> Iterator<ToXContent> map(String name, Map<String, T> map, Fun
         return wrapWithObject(name, Iterators.map(map.entrySet().iterator(), toXContent));
     }
 
+    /**
+     * Creates an Iterator of a single ToXContent object that serializes all the given 'contents' ToXContent objects into a single chunk.
```
Comment (DaveCTurner) [inline server/src/main/java/org/elasticsearch/common/xcontent/ChunkedToXContentHelper.java:86]: :+1:

**Review by DaveCTurner — COMMENTED** [review]
Looks great, just a few nits

**Comment by DaveCTurner** [comment]
> The GitHub issue has `>enhancement` label, but `>non-issue` seems more appropriate? Let me know which is best.

I think it's a genuine improvement, it will save many MBs of heap sending this response in a large cluster, so I'd call it `>enhancement` so it goes in the release notes.

> There's a lot of wiggle room in what to convert from `toXContent` to `toXContentChunked`. I'm not sure what the rules of thumb are, if there are any. It seems a bit haphazard, at least the way I understand it.

The goal is to have O(1) chunk size in the output, so the key rule is that loops should emit at least one chunk per iteration (unless I guess they have fixed and small bounds). In practice that means that most loops have a chunk before (at least to open the array/object) and one after (to end the array/object). Everything else is really just a subjective question of readability - normally IMO it's easier to have a single chunk that contains a bunch of fields and constant-size arrays/objects, but in some situations it might be more readable to split those things up too.

> Regarding `AbstractChunkedSerializingTestCase.assertChunkCount`, I'm a little lost on how to predict number of chunks expected, aside from just hardcoding whatever the number of chunks happens to be.

If there's lots of tests, or they generate random inputs, or the tests change frequently, then it's worth writing down a computation of the chunk count in terms of the object's size (see e.g. `org.elasticsearch.cluster.ClusterStateTests#expectedChunkCount`). But just a fixed number (greater than 1) is really enough for us for now.

**Review by DaveCTurner — APPROVED** [review]
LGTM :+1:

**Review by DiannaHohensee — COMMENTED** [review]
> If there's lots of tests, or they generate random inputs, or the tests change frequently, then it's worth writing down a computation of the chunk count in terms of the object's size (see e.g. org.elasticsearch.cluster.ClusterStateTests#expectedChunkCount). But just a fixed number (greater than 1) is really enough for us for now.

Okay. I might leave it be, then. The number of chunks seems laborious to determine, since this case has a lot of if-statements and function calls down to the actual chunking.

My testing additions (asserts) are mostly arbitrarily placed.

File: server/src/main/java/org/elasticsearch/cluster/routing/allocation/AbstractAllocationDecision.java (hunk 4 above)
Comment (DaveCTurner) [reply server/src/main/java/org/elasticsearch/cluster/routing/allocation/AbstractAllocationDecision.java]: I don't think that'd prevent the `flatMap`/`single` case. That would forbid returning an empty iterator for one of the items over which we're iterating, but that's a legitimate thing to do. See e.g. here:

https://github.com/elastic/elasticsearch/blob/19257125b19ff05a6b014f37693f9417ca579626/server/src/main/java/org/elasticsearch/cluster/metadata/Metadata.java#L1453-L1458

I don't have a great idea for avoiding this apart from catching it in reviews. I imagine we could write a checkstyle rule for it if it was very important to us, but the consequences of missing it are not terribly severe.

File: docs/changelog/99641.yaml [hunk 6]
```diff
@@ -0,0 +1,5 @@
+pr: 99641
+summary: Chunk the cluster allocation explain response
+area: Network
+type: enhancement
+issues: []
```
Comment (DiannaHohensee) [inline docs/changelog/99641.yaml:5]:
```suggestion
issues: [97803]
```

**Pull request merged**